lib/luaxtable.h

```diff
@@ -0,0 +1,34 @@
+/*
+* Copyright (c) 2024 Mohammad Shehar Yaar Tausif <sheharyaar48@gmail.com>
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files (the
+* "Software"), to deal in the Software without restriction, including
+* without limitation the rights to use, copy, modify, merge, publish,
+* distribute, sublicense, and/or sell copies of the Software, and to
+* permit persons to whom the Software is furnished to do so, subject to
+* the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef luaxtable_h
+#define luaxtable_h
+
+struct luaxtable_s;
+
+typedef struct luaxtable_info_s {
+	struct luaxtable_s *data __attribute__((aligned(8)));
+} luaxtable_info_t;
```

Comment:
can't we simply use only one struct for both? why do we need the nested one?

Reply:
> The following are the struct parameters that are used by match and target functions :
> 
> * `match` : struct sk_buff, **const** struct **xt_action_param**
> * `target` : **const** struct sk_buff, struct **xt_action_param**
> * `match_checkentry` : const struct xt_**mtchk**_param
> * `target_checkentry` : const struct xt_**tgchk**_param
> * `match_destroy` : const struct xt_**mtdtor**_param
> * `target_destory` : const struct xt_**tgdtor**_param

The struct `luaxtable_info_s` is set under `params->(target/match)info`. Now since the struct param is const in (match/target) checkentry, we cannot edit the info field. That's why we need another level of indirection (const param -> match/target info -> data), here data can be modified

Reply:
This is the pattern across xtable kernel and userspace addon source codes.
For example : 
`xt_quota.h` : https://elixir.bootlin.com/linux/v6.9/source/include/uapi/linux/netfilter/xt_quota.h#L14
`xt_quota.c` : https://elixir.bootlin.com/linux/v6.9/source/net/netfilter/xt_quota.c#L46

Reply:
got it.. I just don't see the `__attribute__((aligned(8))` in the example you shared.. but agreed, let's use the same pattern..

Reply:
That is given under  `Section 4.5 :  Attaching kernel-specific data` in https://inai.de/documents/Netfilter_Modules.pdf

Other examples use this : 
1. xt_rateest.h : https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/netfilter/xt_rateest.h#L21
2. xt_TEE.h : https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/netfilter/xt_TEE.h#L12
3. xt_cgroup.h : https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/netfilter/xt_cgroup.h#L22

and many more under `include/uapi/linux/netfilter/*.h`